### PR TITLE
Go SDK Combine load tests

### PR DESCRIPTION
**Pull request by tszerszen**
This PR adds Combine load tests for Go SDK.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by tszerszen** [comment]
R: @kamilwu ?

File: .test-infra/jenkins/job_LoadTests_Combine_Flink_Go.groovy [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import InfluxDBCredentialsHelper
+
+import static LoadTestsBuilder.DOCKER_CONTAINER_REGISTRY
+
+String now = new Date().format('MMddHHmmss', TimeZone.getTimeZone('UTC'))
+
+def batchScenarios = {
+  [
+    [
+      title          : 'Group By Key Go Load test: 2GB of 10B records',
```
Comment (kamilwu) [inline .test-infra/jenkins/job_LoadTests_Combine_Flink_Go.groovy:33]: `Group By Key Go Load test` -> `Combine Go Load test`

File: .test-infra/jenkins/job_LoadTests_Combine_Flink_Go.groovy [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        environment_type     : 'DOCKER',
+        environment_config   : "${DOCKER_CONTAINER_REGISTRY}/beam_go_sdk:latest",
+      ]
+    ],
+    [
+      title          : 'Group By Key Go Load test: fanout 4 times with 2GB 10-byte records total',
+      test           : 'combine',
+      runner         : CommonTestProperties.Runner.FLINK,
+      pipelineOptions: [
+        job_name             : "load-tests-go-flink-batch-combine-2-${now}",
```
Comment (kamilwu) [inline .test-infra/jenkins/job_LoadTests_Combine_Flink_Go.groovy:57]: Please replace "2" with "4" here and below, at line 59. This is for compatibility reasons. Some time ago, we had tests 2 and 3, but they were eventually removed.

File: .test-infra/jenkins/job_LoadTests_Combine_Flink_Go.groovy [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        environment_type     : 'DOCKER',
+        environment_config   : "${DOCKER_CONTAINER_REGISTRY}/beam_go_sdk:latest",
+      ]
+    ],
+    [
+      title          : 'Group By Key Go Load test: fanout 8 times with 2GB 10-byte records total',
+      test           : 'combine',
+      runner         : CommonTestProperties.Runner.FLINK,
+      pipelineOptions: [
+        job_name             : "load-tests-go-flink-batch-combine-3-${now}",
```
Comment (kamilwu) [inline .test-infra/jenkins/job_LoadTests_Combine_Flink_Go.groovy:77]: "3" -> "5"

File: .test-infra/jenkins/job_LoadTests_Combine_Flink_Go.groovy [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    'beam_LoadTests_Go_Combine_Flink_Batch',
+    'Run Load Tests Go Combine Flink Batch',
+    'Load Tests Go Combine Batch suite',
+    this
+    ) {
+      additionalPipelineArgs = [:]
+      loadTestJob(delegate, CommonTestProperties.TriggeringContext.PR, 'batch')
+    }
+
+CronJobBuilder.cronJob('beam_LoadTests_Go_Combine_Flink_Batch', 'H 10 * * *', this) {
```
Comment (kamilwu) [inline .test-infra/jenkins/job_LoadTests_Combine_Flink_Go.groovy:130]: We already have two tests firing at the same time (at 10 am UTC): Go ParDo (merged) and Go GBK (will be merged soon). Let's choose something else. I think `8` is a good choice.

File: sdks/go/test/load/combine/combine.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/apache/beam/sdks/go/pkg/beam"
+	"github.com/apache/beam/sdks/go/pkg/beam/io/synthetic"
+	"github.com/apache/beam/sdks/go/pkg/beam/log"
+	"github.com/apache/beam/sdks/go/pkg/beam/transforms/top"
+	"github.com/apache/beam/sdks/go/pkg/beam/x/beamx"
+	"github.com/apache/beam/sdks/go/test/load"
+)
+
+var (
+	fanout          = flag.Int(
```
Comment (kamilwu) [inline sdks/go/test/load/combine/combine.go:32]: This formatting looks a bit strange. Can we do it in the same way as for test GBK test?

File: sdks/go/test/load/combine/combine.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+		"fanout",
+		1,
+		"A number of combine operations to perform in parallel.")
+	topCount        = flag.Int(
+		"top_count",
+		20,
+		"A number passed to the combiner.")
```
Comment (kamilwu) [inline sdks/go/test/load/combine/combine.go:39]: What about something like this: `A number of greatest elements to extract from the PCollection.` ?

File: sdks/go/test/load/combine/combine.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+func parseSyntheticConfig() synthetic.SourceConfig {
+	if *syntheticConfig == "" {
+		panic("--input_options not provided")
+	} else {
+		encoded := []byte(*syntheticConfig)
+		return synthetic.DefaultSourceConfig().BuildFromJSON(encoded)
+	}
+}
+
+func CompareLess(key []byte, value []byte) bool {
```
Comment (kamilwu) [inline sdks/go/test/load/combine/combine.go:55]: In Go, if a name begins with a capital letter, it is an exported name. I think there's no need to mark this function as an exported one.

File: sdks/go/test/load/combine/combine.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	flag.Parse()
+	beam.Init()
+
+	ctx := context.Background()
+
+	p, s := beam.NewPipelineWithRoot()
+	src := synthetic.SourceSingle(s, parseSyntheticConfig())
+	pcoll := beam.ParDo(s, &load.RuntimeMonitor{}, src)
+	for i := 0; i < *fanout; i++ {
+		pcoll = top.LargestPerKey(s, pcoll, *topCount, CompareLess)
```
Comment (kamilwu) [inline sdks/go/test/load/combine/combine.go:69]: Two transforms are missing after the combing operation:
- a ParDo for consuming output elements. Please visit `sdks/python/apache_beam/testing/load_tests/combine_test.py` to check how it was implemented in Python
- a final ParDo with a RuntimeMonitor

**Review by kamilwu — COMMENTED** [review]
Thanks @tszerszen. I left some comments.

File: sdks/go/test/load/combine/combine.go (hunk 8 above)
Comment (kamilwu) [reply sdks/go/test/load/combine/combine.go:69]: Those transform should be put in a for loop, just like in Python.

**Pull request merged**